### PR TITLE
refactor StatelessFuncAnalyzer

### PR DESCRIPTION
**Pull request by omsmith**
image

### CODE REVIEW

File: tests/D2L.CodeStyle.Analyzers.Test/Specs/StatelessFuncAnalyzerSpec.cs [hunk 1]
```diff
@@ -64,20 +65,20 @@ internal sealed class Usages {
...
-		public void DelegateStatic() {
+		public void LambdaStatic() {
 			var func = new StatelessFunc<string, string>( static delegate ( string x ) { return x + "\n"; } );
 			AttributeFuncReceiver.Accept<string, string>( static delegate ( string x ) { return x + "\n"; } );
 		}
 
-		public void DelegateNoClosures() {
-			var func = new StatelessFunc<string, string>( /* StatelessFuncIsnt(Delegate is not static) */ delegate ( string x ) { return x + "\n"; } /**/ );
-			AttributeFuncReceiver.Accept<string, string>( /* StatelessFuncIsnt(Delegate is not static) */ delegate ( string x ) { return x + "\n"; } /**/ );
+		public void LambdaNoClosures() {
```
Comment (omsmith) [inline tests/D2L.CodeStyle.Analyzers.Test/Specs/StatelessFuncAnalyzerSpec.cs:73]: Oops, came from Ctrl+H'ing the error message

File: src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs [hunk 2]
```diff
@@ -27,167 +20,176 @@ public sealed class StatelessFuncAnalyzer : DiagnosticAnalyzer {
...
 					);
 
-					break;
-
+					break;
+
 				/**
-				 * This is the case where a variable is passed in
-				 * eg Foo( f )
-				 * Where f might be a local variable, a parameter, or field
```
Comment (omsmith) [inline src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs:151]: Fields are handled by `IMemberReferenceOperation`

File: src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs [hunk 3]
```diff
@@ -27,167 +20,176 @@ public sealed class StatelessFuncAnalyzer : DiagnosticAnalyzer {
...
+			IOperation operationToInspect = argumentOperation.Value;
+			switch( operationToInspect ) {
+				/**
+				 * Any statement defining an anonymous function:
+				 *   () => 0
+				 *   x => 0
+				 *   delegate () { return 0; }
+				 * We grab the target to get the actual function out instead of
+				 * the act of defining it.
+				 */
```
Comment (j3parker) [inline src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs:72]: nit: we use `//`: https://search.d2l.dev/search?project=Brightspace&full=%22%2F*%22&defs=&refs=&path=%21Specs+%22D2L.CodeStyle.Analyzers%22&hist=&type=&xrd=&nn=1&si=path&si=path

All the extra `*`s you're gonna save bytes anyway :)

File: src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs [hunk 4]
```diff
@@ -27,167 +20,176 @@ public sealed class StatelessFuncAnalyzer : DiagnosticAnalyzer {
...
+				return;
 			}
 
-			ImmutableArray<AttributeData> paramAttributes = param.GetAttributes();
-			if( !paramAttributes.Any( a => a.AttributeClass.Equals( statelessFuncAttribute, SymbolEqualityComparer.Default ) ) ) {
-				return;
+			IParameterSymbol? param = argumentOperation.Parameter;
+			if( param == null ) {
+				return;
 			}
```
Comment (omsmith) [inline src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs:57]: says Parameter can be null for __arglist

**Review by j3parker — APPROVED** [review]

File: src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs (hunk 3 above)
Comment (j3parker) [reply src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs:72]: Lol I see, this file was already doing it. I don't like the inconsistency in this repo (and the rest of our code) but approved regardless.

Comment (omsmith) [reply src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs:72]: Went with what was (mostly) in the file, (which was admittedly me in the first place). Might follow-up.

👍

**Pull request merged**